contrib/core-contract-tests/contracts/pox/pox-4.clar

```diff
@@ -0,0 +1,1355 @@
+                           reward-cycle: uint,
+                           num-cycles: uint,
+                           amount-ustx: uint }))
+  (let ((pox-addr     (get pox-addr     params))
+        (num-cycles   (get num-cycles   params))
+        (reward-cycle (get reward-cycle params))
+        (amount-ustx  (get amount-ustx  params)))
+    (let ((current-amount
+           (default-to u0
+             (get stacked-amount
+                  (map-get? partial-stacked-by-cycle { sender: tx-sender, pox-addr: pox-addr, reward-cycle: reward-cycle })))))
+      (if (>= cycle-index num-cycles)
+          ;; do not add to cycles >= cycle-index
+          false
+          ;; otherwise, add to the partial-stacked-by-cycle
+          (map-set partial-stacked-by-cycle
+                   { sender: tx-sender, pox-addr: pox-addr, reward-cycle: reward-cycle }
+                   { stacked-amount: (+ amount-ustx current-amount) }))
+      ;; produce the next params tuple
+      { pox-addr: pox-addr,
+        reward-cycle: (+ u1 reward-cycle),
+        num-cycles: num-cycles,
+        amount-ustx: amount-ustx })))
+
+;; Add a PoX address to a given sequence of partial reward cycle lists.
+;; A PoX address can be added to at most 12 consecutive cycles.
+;; No checking is done.
+(define-private (add-pox-partial-stacked (pox-addr (tuple (version (buff 1)) (hashbytes (buff 32))))
+                                         (first-reward-cycle uint)
+                                         (num-cycles uint)
+                                         (amount-ustx uint))
+  (let ((cycle-indexes (list u0 u1 u2 u3 u4 u5 u6 u7 u8 u9 u10 u11)))
+    (fold add-pox-partial-stacked-to-ith-cycle cycle-indexes
+          { pox-addr: pox-addr, reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx })
+    true))
+
+;; What is the minimum number of uSTX to be stacked in the given reward cycle?
+;; Used internally by the Stacks node, and visible publicly.
+(define-read-only (get-stacking-minimum)
+    (/ stx-liquid-supply STACKING_THRESHOLD_25))
+
+;; Is the address mode valid for a PoX address?
+(define-read-only (check-pox-addr-version (version (buff 1)))
+    (<= (buff-to-uint-be version) MAX_ADDRESS_VERSION))
+
+;; Is this buffer the right length for the given PoX address?
+(define-read-only (check-pox-addr-hashbytes (version (buff 1)) (hashbytes (buff 32)))
+    (if (<= (buff-to-uint-be version) MAX_ADDRESS_VERSION_BUFF_20)
+        (is-eq (len hashbytes) u20)
+        (if (<= (buff-to-uint-be version) MAX_ADDRESS_VERSION_BUFF_32)
+            (is-eq (len hashbytes) u32)
+            false)))
+
+;; Is the given lock period valid?
+(define-read-only (check-pox-lock-period (lock-period uint))
+    (and (>= lock-period MIN_POX_REWARD_CYCLES)
+         (<= lock-period MAX_POX_REWARD_CYCLES)))
+
+;; Is the current burn block height in the prepare phase?
+(define-read-only (check-not-prepare-phase)
```

Comment:
A few things:

* Can this be an affirmative test?  e.g. `check-prepare-phase`?  Then, the caller performs the negation.
* Instead of relying on globals (e.g. `burn-block-height`), can this instead be an argument?  The caller would supply the global, and the test code would supply arbitrary values without having to set up a whole blockchain to test it.